File: FunctionalPatterns/Creational/Singleton.fs
namespace FunctionalPatterns
namespace FunctionalPatterns.Creational

module Singleton =
    open System

    // Not really a functional solution as we use mutable.
    // Singleton is very difficult (not possible?) to implement purely functional.
    // So here we make use of F# multiparadigm features.
    let mutable private configValue = lazy None

    let private parseInt () =
        Console.WriteLine("Initialize Singleton: Please config a value (int):")
        let line = Console.ReadLine()
        try
           Int32.Parse(line)
        with
           | _ -> 0

    let getValue () =
        match configValue.Force() with
        | Some value -> value
        | None ->
            let value = parseInt ()
            configValue <- lazy Some(value)
            value

File: ObjectOrientedPatterns/Creational/Singleton.cs
namespace ObjectOrientedPatterns.Creational.Singleton
{
    // see https://en.wikipedia.org/wiki/Singleton_pattern

    public class ConfigSingleton
    {
        public static ConfigSingleton GetInstance
        {
            get
            {
                instance ??= new ConfigSingleton();
                return instance;
            }
        }
        public int GetValue() => myValue;
        private ConfigSingleton()
        {
            Console.WriteLine("Initialize Singleton: Please config a value (int):");
            string? line = Console.ReadLine();
            if (line != null)
            {
                try
                {
                    myValue = int.Parse(line);
                } catch
                {
                    Console.WriteLine("No parse to int, using 0 as default");
                    myValue = 0;
                }
            }
        }
        private static ConfigSingleton? instance = null;
        private readonly int myValue;
    }
}
